### PR TITLE
daf_persistence - Tickets/dm 11407

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: python/lsst/daf/persistence/posixStorage.py [hunk 1]
```diff
@@ -615,5 +540,248 @@ def storageExists(uri):
...
+    additionalData = butlerLocation.getAdditionalData()
+    locations = butlerLocation.getLocations()
+    with SafeFilename(os.path.join(butlerLocation.getStorage().root, locations[0])) as locationString:
+        logLoc = LogicalLocation(locationString, additionalData)
+        flags = additionalData.getInt("flags", 0)
+        obj.writeFits(logLoc.locString(), flags=flags)
+        return
+
+
+def readPafStorage(self, butlerLocation):
```
Comment (andy-slac) [inline python/lsst/daf/persistence/posixStorage.py]: `self`?

File: python/lsst/daf/persistence/posixStorage.py [hunk 2]
```diff
@@ -615,5 +540,248 @@ def storageExists(uri):
...
+    Returns
+    -------
+    A list of objects as described by the butler location. One item for
+    each location in butlerLocation.getLocations()
+    """
+    results = []
+    for locationString in butlerLocation.getLocations():
+        logLoc = LogicalLocation(butlerLocation.getStorage().locationWithRoot(locationString),
+                                 butlerLocation.getAdditionalData())
+        finalItem = Policy(filePath=logLoc.locString())
```
Comment (andy-slac) [inline python/lsst/daf/persistence/posixStorage.py]: I guess this Policy is different from pexPolicy? Sort of confusing.

File: python/lsst/daf/persistence/storageInterface.py [hunk 3]
```diff
@@ -64,28 +65,69 @@ def registerFormatter(cls, formatable, formatter):
             The type of StorageInterface the formatter is being registered for.
         formatable : class object
             The class object whose instances can be formatted by the formatter.
-        formatter : Formatter class
-            The formatter class that can be used by the StorageInterface instance to read and write the object
-            to the storage.
+        readFormatter : a read formatter callable
+            The formatter class that can be used by the StorageInterface instance to read the object from the
```
Comment (andy-slac) [inline python/lsst/daf/persistence/storageInterface.py]: `formatter class` - > `method`? (or `function` if you prefer)

File: python/lsst/daf/persistence/storageInterface.py [hunk 4]
```diff
@@ -64,28 +65,69 @@ def registerFormatter(cls, formatable, formatter):
...
+        cls.registerReadFormatter(formatable, readFormatter)
+        cls.registerWriteFormatter(formatable, writeFormatter)
+
+    @classmethod
+    def registerReadFormatter(cls, formatable, readFormatter):
+        """Similar to registerFormatters, but only for the readFormatter."""
+        cls._registerFormatter(formatable, readFormatter, 'read')
+
+    @classmethod
+    def registerWriteFormatter(cls, formatable, writeFormatter):
```
Comment (andy-slac) [inline python/lsst/daf/persistence/storageInterface.py]: Instead three separate methods `registerFormatters`, `registerReadFormatter`, `registerWriteFormatter` I'd probably define just one with keyword arguments:

    def registerFormatters(cls, formattable, read=None, write=None)
        ....

Feels more Pythonic to me :snake:

File: python/lsst/daf/persistence/storageInterface.py [hunk 5]
```diff
@@ -64,28 +65,69 @@ def registerFormatter(cls, formatable, formatter):
...
+        ----------
+        objType : class type
+            The type of class to find a formatter for.
+
+        Returns
+        -------
+        formatter callable
+            The formatter callable used to read the object from the storageInterface.
+        """
+        classFormatters = StorageInterface.readFormatters.get(cls, None)
```
Comment (andy-slac) [inline python/lsst/daf/persistence/storageInterface.py]: One minor issue with using `cls` as a key in a global map — if, for example, I decide to implement new Storage as subclass of PosixStorage (not sure if that is possible) then calling `self._getFormatter()` in a subclass will not find any formatters registered with PosixStorage?

File: python/lsst/daf/persistence/fmtPosixRepositoryCfg.py [hunk 6]
```diff
@@ -22,100 +22,95 @@
...
-        A RepositoryCfg instance or None
-        """
-        repositoryCfg = yaml.load(fileObject)
-        if repositoryCfg is not None:
-            if repositoryCfg.root is None:
-                repositoryCfg.root = uri
-        return repositoryCfg
+    Returns
+    -------
+    A RepositoryCfg instance or None
```
[the remark this answers is not in the document]
Comment (n8pease) [reply python/lsst/daf/persistence/fmtPosixRepositoryCfg.py]: yes

File: python/lsst/daf/persistence/posixStorage.py [hunk 7]
```diff
@@ -615,5 +540,248 @@ def storageExists(uri):
...
+    results = []
+    additionalData = butlerLocation.getAdditionalData()
+    for locationString in butlerLocation.getLocations():
+        locStringWithRoot = os.path.join(butlerLocation.getStorage().root, locationString)
+        logLoc = LogicalLocation(locStringWithRoot, additionalData)
+        if not os.path.exists(logLoc.locString()):
+            raise RuntimeError("No such pickle file: " + logLoc.locString())
+        with open(logLoc.locString(), "rb") as infile:
+            # py3: We have to specify encoding since some files were written
+            # by python2, and 'latin1' manages that conversion safely. See:
```
[the remark this answers is not in the document]
Comment (n8pease) [reply python/lsst/daf/persistence/posixStorage.py]: #notmycode #movedfromelsewhere

File: python/lsst/daf/persistence/posixStorage.py (hunk 2 above)
Comment (n8pease) [reply python/lsst/daf/persistence/posixStorage.py]: yeah. naming is hard. sorry.

File: python/lsst/daf/persistence/storageInterface.py (hunk 5 above)
Comment (n8pease) [reply python/lsst/daf/persistence/storageInterface.py]: yeah, it's broken. I ran into it the other day. Needs fixing.

**Pull request merged**